### PR TITLE
Implement the Settings UI

### LINKED ISSUE #1564: Feature Request - Implement Settings UI
# Summary of the new feature/enhancement

Implement a Settings UI for Windows Terminal. This UI should read from and update the settings.json file. The user should still be able to edit the JSON directly, if so desired.

Spec & Design: #6720

# Implementation Plan
By @carlos-zamora. We're working off of the `feature/settings-ui` branch. It runs the same CI as `master`. The idea is that you make PRs targeting `feature/settings-ui` and when the Settings UI is ready for delivery, we merge it into `master`.

We're basically starting with #7370.

## Back-end
These work items generally won't touch _any_ UI or XAML.

- [X] #885 - TerminalSettingsModel
   - This is needed to be able to access the contents of the settings file. The settings objects must be COM objects. They can then be exposed to a Settings UI to read and modify Windows Terminal's settings. (PR #7667, #7830)
   - Spec: #6904
- [x] Write settings to JSON
   - [x] **Part 1**: JsonUtils includes `ConversionTrait`, which is used to convert settings from JSON to our settings objects (`FromJson()`). We should be able to introduce a `ToJson()` function to `ConversionTrait`. This will mainly handle individual settings being converted to their enum value. (PR #7609)
   - [x] **Part 2**: Introduce `ToJson()`  as a function on full settings objects (i.e: Profile, GlobalAppSettings, etc...) (PR #8018)
   - [ ] **Part 3**: Layer the output JSON from part 2, onto the user's existing settings. We need to make sure we don't overwrite any comments placed by the user. 
- [x] Update `openSettings` command to open the Settings UI
   - Spec: #5915
   - PR: #7802

## Middle-end
These work items may touch a bit of the XAML. You'll have to touch the `TerminalSettingsEditor` project, but you won't be too caught up in the polish of icons, margins, etc.

- [x] Embed the Settings UI as a tab in Windows Terminal (TerminalApp)
   - Convert `TerminalSettingsEditor` into a DLL
   - Throw that `TerminalSettingsEditor` into a tab for Terminal
   - Bonus Points: we should probably only have one Settings UI open at a time
   - PRs: #7675, #7802
- [x] Replace `ObjectModel` objects in the prototype with real `TerminalSettingsModel` object references
- [x] Data Bind the settings to the `TerminalSettingsModel`
   - Global Settings
   - Profiles:
      - List of profiles --> Navigation bar
      - Profile settings
   - Color Schemes:
      - List of color schemes --> dropdown menu
      - colors
   - Key Bindings

#### Miscellaneous Follow-Ups from PRs
- [x] Use the same UI for a new profile and existing profile.
- [x] `ColorSchemes.xaml` button color flyouts should be all in a single control.
- [x] Ensure all strings are attached to UIDs. (PR #7833 )
- [x] Commandline handling needs to be robust and intelligent.
- [x] Push localizable strings into a `Resources.resw` (PR #7833)
- [x] Dedupe `Utils`'s `GetFontIcon` function with `Command`'s magic icon handler as per #7802
- [X] Hide warning, and change save/reset buttons to apply/discard (PR #8082 )
   - we don't have a system to detect changes yet, so this'll do in the meantime.

## Front-end
These work items include a bit more design work. Make sure you look at the mockups in #6720.

- [X] Update icons (PR #8013)
	- Profiles: E756
	- Appearance: E771
	- Color Schemes: E790
	- Startup: E7B5
- [X] Hook up button to open the JSON settings (see #5915) (PR #8059)
- [x] Add and hook up "Save" button

## Bonus Points
- [ ] Search for settings
- [ ] Home Page
- [ ] Image preview for icons and background image
- [ ] Embed a Terminal Control in the "Appearance" page for profiles
   - This should live update with changes you make to colors, background image, font, etc.
   - Requires a new Terminal Connection specifically designed for this (or even throw a fully operational terminal in there, sure!)
- [ ] Tab title changes depending on what SUI page is currently open.
- [ ] Implement the Keyboard page (key bindings)
   - See #6720 for more details

## Bugs
These are bugs that the team is finding as we selfhost the feature branch. This method will be used until the feature branch gets merged into master. At that point, we'll submit bugs normally.

- [x] TSM and TSE localization not appearing
  - More info: Build authoring error: add `<Import Project="$(SolutionDir)build\rules\CollectWildcardResources.targets" />` to TSM and TSE vcxproj (PR #7966 #7967 )
- [x] The "unsaved changes" thing obstructs the settings at the bottom of the page, and they can't be scrolled up
- [ ] Unlocalized strings:
  - [X] "Settings" on `SettingsTab` (PR #8012 )
  - [ ] "Global" in UI Nav Menu (under Profiles)
    - This string already has a uuid and is defined in the resw. Need to explore deeper.
- [X] If string for save/reset button is too long, it is cropped (PR #8082 )
- [X] Spacing is incorrect on the Profiles - Global - Advanced page (PR #8065)
- [x] Hide keyboard page for 1.5 release (PR #8078)
- [x] Hide home page for 1.5 release (PR #8078 )

# Designs
By @cinnamon-msft. Content within these designs is not final and will be modified. They're simply here for demonstration.

navigation-2

appearance

### PR DESCRIPTION
**Pull request by leonMSFT**
## Summary of the Pull Request
This introduces the Settings UI to Windows Terminal. This allows users to customize their global settings, color schemes, and profiles without having to open the settings.json. Actions are not currently included in the Settings UI.

To open the Settings UI, add the following keybinding to your settings.json:
```js
{ "name": "OPEN SETTINGS UI", "command": {"action": "openSettings", "target": "settingsUI"}, "keys": "ctrl+shift+," }
```

Closes #1564 - Settings UI
Spec: #6720, #8269

## Demo
Settings UI Demo

### CODE REVIEW

**Review by miniksa — APPROVED** [review]
Only minor comments. Let's do this!

File: src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+    <value>When checked, a selection is immediately copied to your clipboard upon creation. When unchecked, the selection persists and awaits further action.</value>
+  </data>
+  <data name="Globals_DefaultProfile.Header" xml:space="preserve">
+    <value>Default profile</value>
+  </data>
+  <data name="Globals_DefaultProfile.[using:Windows.UI.Xaml.Controls]ToolTipService.ToolTip" xml:space="preserve">
+    <value>Profile that opens when clicking the '+' icon or by typing the key binding assigned to 'newTab'.</value>
+  </data>
+  <data name="Globals_DisableDynamicProfiles.Content" xml:space="preserve">
+    <value>Automatically create new profiles when new shells are installed</value>
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Resources/en-US/Resources.resw:212]: This setting isn't even exposed right now. So I'm removing it entirely.

**Comment by zadjii-msft** [comment]
Random thought while reading: we might want to actually make an updated version of `AddASetting.md` for the _Terminal_ now with all this. For ex, if you add any sort of enum setting, you need to add it to a _bunch_ of different places so it can be exposed in the SUI. Maybe a checklist would be helpful to have?

File: src/cascadia/TerminalSettingsEditor/GlobalSettings.idl [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/GlobalSettings.idl:1]: Uh, should this file be kill?

File: src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.filters [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    </ClCompile>
+    <ClCompile Include="ObjectModel\GlobalSettings.cpp">
+      <Filter>ObjectModel</Filter>
+    </ClCompile>
+    <ClCompile Include="ObjectModel\GlobalSettingsModel.cpp">
+      <Filter>ObjectModel</Filter>
+    </ClCompile>
+    <ClCompile Include="ObjectModel\Profile.cpp">
+      <Filter>ObjectModel</Filter>
+    </ClCompile>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.filters:25]: These probably need to get removed too, yea? (and below)

File: src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.filters [hunk 4]
```diff
@@ -0,0 +1,100 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/Microsoft.Terminal.Settings.Editor.filters:1]: wait is this file needed at all anymore?

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
okay I'm at 69/90 but I think he's finally asleep so I'm going to bed

File: src/cascadia/TerminalApp/SettingsTab.cpp [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            auto fontFamily = winrt::WUX::Media::FontFamily(L"Segoe MDL2 Assets");
+            auto glyph = L"\xE713"; // This is the Setting icon (looks like a gear)
+
+            // The TabViewItem Icon needs MUX while the IconSourceElement in the CommandPalette needs WUX...
+            Icon(glyph);
+            TabViewItem().IconSource(IconPathConverter::IconSourceMUX(glyph));
+
+            // Update SwitchToTab command's icon
+            SwitchToTabCommand().Icon(glyph);
```
[the remark this answers is not in the document]
Comment (Don-Vito) [reply src/cascadia/TerminalApp/SettingsTab.cpp:92]: @zadjii-msft - not sure about the git but the compiler will complain for sure :) In any case the solution will be straightforwad - removing this line since the palette items are "bound" to `TabBase::Icon`.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    }
+
+    void MainPage::OpenJsonKeyDown(IInspectable const& /*sender*/, Windows::UI::Xaml::Input::KeyRoutedEventArgs const& args)
+    {
+        if (args.Key() == VirtualKey::Enter || args.Key() == VirtualKey::Space)
+        {
+            const auto target = args.KeyStatus().IsMenuKeyDown ? SettingsTarget::DefaultsFile : SettingsTarget::SettingsFile;
+            _OpenJsonHandlers(nullptr, target);
+        }
+    }
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/MainPage.cpp:240]: Wait don't buttons just handle a Space or Enter themselves? I didn't think you had to do this manually

File: src/cascadia/TerminalSettingsEditor/ColorToBrushConverter.cpp [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+using namespace winrt::Windows::UI::Xaml;
+
+namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
+{
+    Foundation::IInspectable ColorToBrushConverter::Convert(Foundation::IInspectable const& value,
+                                                            Windows::UI::Xaml::Interop::TypeName const& /* targetType */,
+                                                            Foundation::IInspectable const& /* parameter */,
+                                                            hstring const& /* language */)
+    {
+        return winrt::box_value(Windows::UI::Xaml::Media::SolidColorBrush(winrt::unbox_value<Windows::UI::Color>(value)));
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/ColorToBrushConverter.cpp:18]: _Technically_, yes. But considering that this is a Color-to-Brush converter, I feel like this is fine? (also, idk how to check the type of something without unboxing it, would `try_as` work?)

File: src/cascadia/TerminalSettingsEditor/Utils.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        if (key >= VirtualKey::NumberPad0 && key <= VirtualKey::NumberPad9)
+        {
+            return L"numpad_" + hstring(std::to_wstring((int)key - (int)VirtualKey::NumberPad0));
+        }
+
+        switch (key)
+        {
+        case VirtualKey::Control:
+            return L"ctrl+";
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Utils.cpp:56]: Turns out we don't actually need this function. It's not used anywhere. Must've been left over from the keybindings page.

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 9]
```diff
@@ -2741,6 +2753,73 @@ namespace winrt::TerminalApp::implementation
...
+
+            // Give the tab its index in the _tabs vector so it can manage its own SwitchToTab command.
+            _UpdateTabIndices();
+
+            // Don't capture a strong ref to the tab. If the tab is removed as this
+            // is called, we don't really care anymore about handling the event.
+            auto weakTab = make_weak(newTabImpl);
+
+            auto tabViewItem = newTabImpl->TabViewItem();
+            _tabView.TabItems().Append(tabViewItem);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2798]: A lot of this seems similar to the setup we do when we create a terminal tab, should we make a card to unify it?

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 10]
```diff
@@ -2741,6 +2753,73 @@ namespace winrt::TerminalApp::implementation
...
+            // When the tab is closed, remove it from our list of tabs.
+            newTabImpl->Closed([tabViewItem, weakThis{ get_weak() }](auto&& /*s*/, auto&& /*e*/) {
+                if (auto page{ weakThis.get() })
+                {
+                    page->_switchToSettingsCommand = nullptr;
+                    page->_RemoveOnCloseRoutine(tabViewItem, page);
+                }
+            });
+
+            _switchToSettingsCommand = newTabImpl->SwitchToTabCommand();
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2811]: If the tabs get re-ordered, does the `_switchToSettingsCommand` get updated? Or does it still hold the old value from pre-reorder

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Okay, I'm through 90/90. I've got a pile of nits and questions so I'll hold for answers, but there's nothing _dramatic_ here

File: src/cascadia/TerminalSettingsEditor/ColorToBrushConverter.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/ColorToBrushConverter.cpp:18]: yea, `try_as` returns null if it couldn't do the conversion.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    fire_and_forget Profiles::BackgroundImage_Click(IInspectable const&, RoutedEventArgs const&)
+    {
+        auto lifetime = get_strong();
+
+        FileOpenPicker picker;
+
+        _State.WindowRoot().TryPropagateHostingWindow(picker); // if we don't do this, there's no HWND for it to attach to
+        picker.ViewMode(PickerViewMode::Thumbnail);
+        picker.SuggestedStartLocation(PickerLocationId::PicturesLibrary);
+        picker.FileTypeFilter().ReplaceAll({ L".jpg", L".jpeg", L".png", L".gif" });
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:98]: Just updated the list according to [BitmapImage docs](https://docs.microsoft.com/en-us/uwp/api/Windows.UI.Xaml.Media.Imaging.BitmapImage?view=winrt-19041#remarks) since we convert the URI into a BitmapImage down in TermControl.

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        StorageFile file = co_await picker.PickSingleFileAsync();
+        if (file != nullptr)
+        {
+            Commandline().Text(file.Path());
+        }
+    }
+
+    fire_and_forget Profiles::StartingDirectory_Click(IInspectable const&, RoutedEventArgs const&)
+    {
+        auto lifetime = get_strong();
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:127]: I'll be honest, I straight up took this from the [FilePicker Sample](https://github.com/microsoft/Windows-universal-samples/blob/master/Samples/FilePicker/cppwinrt/Scenario1_SingleFile.cpp). We don't even use `lifetime`. I'll have to play around a bit here to see if that works. I'll get back to you.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp (hunk 6 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:240]: _Surprisingly_, this is also necessary. Removing the KeyDown handler _completely_ removed functionality for invoking it with the keyboard. Adding a comment in the .xaml file.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 10 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2811]: Just tested it. It works!

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 9 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2798]: This was some of the code that Leon wrote up. I'm a bit worried that I may break some tab switching functionality in the process of moving stuff around. For now, I've lumped it into #8452 since that seems to be touching some similar code. I'll do it as a follow-up to this PR since this is already a pretty big one.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 13]
```diff
@@ -55,29 +55,28 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
-                    continue;
-                }
-                else if (const auto stringTag{ tag.try_as<hstring>() })
+                if (const auto tag{ navViewItem.Tag() })
                 {
-                    if (stringTag == addProfileTag)
+                    if (tag.try_as<Model::Profile>())
+                    {
+                        // remove NavViewItem pointing to a Profile
+                        navViewItem.Visibility(Visibility::Collapsed);
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/MainPage.cpp]: come now, you've _gotta_ put a huge comment above code that looks weird like this

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp [hunk 14]
```diff
@@ -104,6 +107,27 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         }
     }
 
+    fire_and_forget Profiles::Icon_Click(IInspectable const&, RoutedEventArgs const&)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/Profiles.cpp:110]: we should file a follow-up to de-dupe these two. The only difference is which property ends up getting set

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:248]: w a t. Okay, thanks for the follow up.

File: src/cascadia/TerminalSettingsEditor/MainPage.xaml [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+            </Grid.RowDefinitions>
+            <Frame x:Name="contentFrame"
+                   Grid.Row="0"></Frame>
+            <Grid Height="100" Grid.Row="1"
+                  BorderBrush="{ThemeResource SystemBaseLowColor}"
+                  BorderThickness="0,1,0,0">
+                <Grid.ColumnDefinitions>
+                    <ColumnDefinition Width="*"></ColumnDefinition>
+                </Grid.ColumnDefinitions>
+                <TextBlock x:Uid="Settings_UnsavedSettingsWarning"
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsEditor/MainPage.xaml:120]: This currently isnt used, right?

**Review by zadjii-msft — APPROVED** [review]
okay. lets do this. We can polish in post

File: src/cascadia/TerminalSettingsEditor/MainPage.xaml (hunk 15 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsEditor/MainPage.xaml:120]: Correct. We need a way to detect unsaved changes to properly show/hide it. #8552 starts getting us there, so I'll keep it in there for now. But I'll definitely add a comment explaining that (and filing a follow-up)

File: src/cascadia/TerminalSettingsEditor/Profiles.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/Profiles.cpp:151]: so, it really doesn't matter. `lifetime` anchors the page so that it stays alive while the button click handler is being run.

That's _a good thing_.

We've gone crazy with the weak references in Terminal. We make them weak, check if they're alive, all the dang time

File: src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h [hunk 16]
```diff
@@ -168,7 +167,7 @@ struct ::Microsoft::Terminal::Settings::Model::JsonUtils::ConversionTrait<::winr
         }
 
         ::winrt::Windows::UI::Text::FontWeight weight{
-            static_cast<uint16_t>(std::clamp(value, 100u, 990u))
+            static_cast<uint16_t>(std::clamp(value, 1u, 999u))
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:170]: I'm sorry, what? 1 is not a valid font weight

Comment (mdtauk) [reply src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:170]: > I'm sorry, what? 1 is not a valid font weight

Typically goes from 100 to 990 I think - but I believe the font sets its minimum and maximum values?

https://docs.microsoft.com/en-us/typography/opentype/spec/fvar

``
Example
This example is for a hypothetical font with family name “SelawikV” that has two axes of variation, for weight and width. This table summarizes the description of the axes for the font:
``
Axis tag | Minimum value | Default value | Maximum value | Axis name ID
-- | -- | -- | -- | --
'wght' | 300 | 400 | 700 | 256
'wdth' | 62.5 | 100 | 150 | 257

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h]: But the documentation for [FontWeight](https://docs.microsoft.com/en-us/uwp/api/windows.ui.text.fontweight?view=winrt-19041) says that this is the supported range though. Do we clamp it somewhere else?

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h]: I backed this out and clamped it to 100 on deserialization. On a round trip it will be "fixed".
